# Neural Networks

This week we're going to do some simple exploration of the feature space of Neural Networks.

In [1]:
# Building On Example from https://machinelearningmastery.com/tensorflow-tutorial-deep-learning-with-tf-keras/
# Building On Example from https://www.tensorflow.org/tutorials/images/cnn
# Building On Example from https://www.guru99.com/rnn-tutorial.html
# Building On Example from https://machinelearningmastery.com/prepare-univariate-time-series-data-long-short-term-memory-networks/
# Building On Example from TensorFlow Documentation
# For a multi-feature example, see: https://github.com/Hvass-Labs/TensorFlow-Tutorials/blob/master/23_Time-Series-Prediction.ipynb


%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
import sklearn.datasets
import sklearn.model_selection

2023-04-04 00:33:09.514423: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-04-04 00:33:09.623490: I tensorflow/core/util/util.cc:169] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.


First, let's set up a dataset that should be simple enough for us to solve, but complex enough for a variety of Neural Network tasks, including CNNs. We know digits already, and it's a good, usable set of images, so let's start there.

In [2]:
images = sklearn.datasets.load_digits()
X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(images.data, images.target, test_size=0.2, random_state=42)
print(X_train.shape, X_test.shape, y_train, y_test)

(1437, 64) (360, 64) [6 0 0 ... 2 7 1] [6 9 3 7 2 1 5 2 5 2 1 9 4 0 4 2 3 7 8 8 4 3 9 7 5 6 3 5 6 3 4 9 1 4 4 6 9
 4 7 6 6 9 1 3 6 1 3 0 6 5 5 1 9 5 6 0 9 0 0 1 0 4 5 2 4 5 7 0 7 5 9 5 5 4
 7 0 4 5 5 9 9 0 2 3 8 0 6 4 4 9 1 2 8 3 5 2 9 0 4 4 4 3 5 3 1 3 5 9 4 2 7
 7 4 4 1 9 2 7 8 7 2 6 9 4 0 7 2 7 5 8 7 5 7 7 0 6 6 4 2 8 0 9 4 6 9 9 6 9
 0 3 5 6 6 0 6 4 3 9 3 9 7 2 9 0 4 5 3 6 5 9 9 8 4 2 1 3 7 7 2 2 3 9 8 0 3
 2 2 5 6 9 9 4 1 5 4 2 3 6 4 8 5 9 5 7 8 9 4 8 1 5 4 4 9 6 1 8 6 0 4 5 2 7
 4 6 4 5 6 0 3 2 3 6 7 1 5 1 4 7 6 8 8 5 5 1 6 2 8 8 9 9 7 6 2 2 2 3 4 8 8
 3 6 0 9 7 7 0 1 0 4 5 1 5 3 6 0 4 1 0 0 3 6 5 9 7 3 5 5 9 9 8 5 3 3 2 0 5
 8 3 4 0 2 4 6 4 3 4 5 0 5 2 1 3 1 4 1 1 7 0 1 5 2 1 2 8 7 0 6 4 8 8 5 1 8
 4 5 8 7 9 8 5 0 6 2 0 7 9 8 9 5 2 7 7 1 8 7 4 3 8 3 5]


Next, let's make a Feed-Forward Neural Network. We're using the TensorFlow package[https://www.tensorflow.org/learn] to allow us to build these networks by components. See below for the description of each component.

In [3]:
hidden_layers_count = 1 #default 1
hidden_nodes_count = 10 #default 10
hidden_activations = 'relu' #default 'relu'
epochs = 10 #default 10

FFmodel = keras.Sequential()
FFmodel.add(layers.Dense(hidden_nodes_count, input_shape = (64,), activation=hidden_activations)) #Need at least 1 hidden layer
for i in range(hidden_layers_count - 1):
        FFmodel.add(layers.Dense(hidden_nodes_count, activation=hidden_activations))
FFmodel.add(layers.Dense(10, activation='softmax')) #Our output layer

FFmodel.compile(optimizer='sgd', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
FFmodel.fit(X_train, y_train, epochs=epochs, batch_size=32) 


Epoch 1/10


2023-04-03 16:16:05.357470: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


45/45 [==============================] - 0s 1ms/step - loss: 2.7659 - accuracy: 0.1761
Epoch 2/10
45/45 [==============================] - 0s 979us/step - loss: 1.9368 - accuracy: 0.2804
Epoch 3/10
45/45 [==============================] - 0s 933us/step - loss: 1.7105 - accuracy: 0.3535
Epoch 4/10
45/45 [==============================] - 0s 923us/step - loss: 1.4320 - accuracy: 0.4906
Epoch 5/10
45/45 [==============================] - 0s 922us/step - loss: 1.1489 - accuracy: 0.5797
Epoch 6/10
45/45 [==============================] - 0s 888us/step - loss: 0.9310 - accuracy: 0.6729
Epoch 7/10
45/45 [==============================] - 0s 927us/step - loss: 0.7894 - accuracy: 0.7223
Epoch 8/10
45/45 [==============================] - 0s 942us/step - loss: 0.6817 - accuracy: 0.7794
Epoch 9/10
45/45 [==============================] - 0s 921us/step - loss: 0.5809 - accuracy: 0.8051
Epoch 10/10
45/45 [==============================] - 0s 913us/step - loss: 0.5088 - accuracy: 0.8323


"Sequential" is the object that allows us to add layers in sequence (hence the name) to build our model piece by piece.

"Dense" is a fully connected layer of nodes. I set how many nodes I want in the layer and the activation function. In the starting layer, I also have to specify the size and shape of my input. In future layers, whatever is output from the previous layer is used as the input to this next layer.

We compile a model to set up the metrics we will use in the fitting process as well as any evaluation metrics.

We fit on the training set.

### Q1: With the default values for number of (1) Hidden Layers, (10) Nodes per layer, ('relu') Activation, and (10) epochs, what do you get for Accuracy? 
#### Q1A: Run it a few times, at least three. What is your highest accuracy and your lowest accuracy after each set of 10 epochs?
*Here you should be able to see whether this process is fully deterministic and always fits to the same results, or has some sort of randomness in it. Does this match your expectation for fitting a NN?*
#### Q1B: What's the trend in your accuracy here?
*Here it should match to your understanding of achieving convergence. Have we converged to an optimal set of internal weights yet?*

a. Highest: .9074
Lowest: .4342

This matches my expectation because the process is not fully deterministic.

b. The accuracy increases each epoch, which means we haven't converged to an optimal set of weights yet.


Now, let's evaluate on the test set.

In [58]:
loss, acc = FFmodel.evaluate(X_test, y_test)
print('Test Accuracy FFNN ' + str(acc))

12/12 [==============================] - 0s 2ms/step - loss: 0.0629 - accuracy: 0.9806
Test Accuracy FFNN 0.980555534362793


### Q2: What is your test accuracy? How does it compare to your training accuracy?
*This should be connecting to your understanding of training goodness and testing goodness, and the implications of underfitting and overfitting.*

The test accuracy is 0.767. This is better than the training accuracy of 0.752, so we can say we're not overfitting the training data, however it's possible we're underfitting because the training accuracy is usually a little higher than the test accuracy when the model is sufficiently fitting the training data.

### Q3: Tune your hyperparameters (number of hidden layers, number of nodes per layer, activation function, epochs) to achieve 100% Training Accuracy. Find *at least two combinations* of hyperparameters that achieve this objective. List the hyperparameter values and Test Accuracy results below, replacing the defaults.

*Q3 Responses Here*

__FIRST FITTED MODEL__
 - hidden_layers_count = 1
 - hidden_nodes_count = 10000
 - hidden_activations = 'relu'
 - epochs = 25
 - Test Accuracy = 0.9806

__SECOND FITTED MODEL__
 - hidden_layers_count = 1
 - hidden_nodes_count = 100
 - hidden_activations = 'elu'
 - epochs = 60 
 - Test Accuracy = 0.9806

### Q4: What do your results above tell you about fitting a Feed-Forward Neural Network? For example...
#### What relationships did you see between hyperparameters?
#### What impacts did hyperparameters have on resulting overfitting or underfitting?
#### Which hyperparameters made the model take longer to train?

Increasing the number of hidden layers decreased the accuracy, while increasing the number of epochs and hidden nodes increased the accuracy. The elu and linear values for the hidden activations parameter resulted in the highest accuracy. More epochs fundementally means more overfitting. Increasing the number epochs and the number of hidden nodes, especially once I tried values on the order of $10^6$, made the model take longer to train.

## BONUS 30 Points

Now, let's try a Convolutional Neural Net! 

_hint: You will need to do the following steps:
 - Reshape your X_train and X_test samples into 2d arrays that are 8\*8 using tensorflow's reshape.
 - add a 2D convolution layer for our input using the keras layer's Conv2D() object.
 - Add a 2D pooling layer for the result of our convolution using something like the keras layer's MaxPooling2D() object.
 - Flatten your layer so it is no longer a bunch of convolved, pooled 2D representations but instead is 1D to fit in a dense layer.
 - End the model in the same way as the FC network above.

### Q 5: What hyperparameters did you use to get 100% accuracy on the training dataset?
 - Description of your 2D Convolution layer - Size of Convolution overlay? Stride Length? Activation?
 - Description of your Pooling layer - size?
 - Description of the number of Convolution + Pooling steps you added.
 - Description of the Dense layers you added after flattening.
 - Description of how many epochs it took to fit.

In [3]:
#A Cell for building a CNN model
# Reshape input data
X_train = tf.reshape(X_train, [-1, 8, 8, 1])
X_test = tf.reshape(X_test, [-1, 8, 8, 1])

# Build the model
model = keras.Sequential([
    keras.layers.Conv2D(filters=16, kernel_size=(3,3), activation='relu', input_shape=(8, 8, 1)),
    keras.layers.MaxPooling2D(pool_size=(2,2)),
    keras.layers.Flatten(),
    keras.layers.Dense(units=128, activation='relu'),
    keras.layers.Dropout(rate=0.2),
    keras.layers.Dense(units=10, activation='softmax')
])

# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
history = model.fit(X_train, y_train, epochs=10, validation_data=(X_test, y_test))


Epoch 1/10


2023-04-04 01:35:25.174832: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


45/45 [==============================] - 1s 8ms/step - loss: 2.0401 - accuracy: 0.4036 - val_loss: 0.6737 - val_accuracy: 0.8222
Epoch 2/10
45/45 [==============================] - 0s 4ms/step - loss: 0.6023 - accuracy: 0.8135 - val_loss: 0.3011 - val_accuracy: 0.9111
Epoch 3/10
45/45 [==============================] - 0s 4ms/step - loss: 0.3095 - accuracy: 0.9151 - val_loss: 0.1848 - val_accuracy: 0.9500
Epoch 4/10
45/45 [==============================] - 0s 4ms/step - loss: 0.2283 - accuracy: 0.9374 - val_loss: 0.1481 - val_accuracy: 0.9611
Epoch 5/10
45/45 [==============================] - 0s 4ms/step - loss: 0.1826 - accuracy: 0.9478 - val_loss: 0.1301 - val_accuracy: 0.9583
Epoch 6/10
45/45 [==============================] - 0s 4ms/step - loss: 0.1351 - accuracy: 0.9645 - val_loss: 0.0936 - val_accuracy: 0.9750
Epoch 7/10
45/45 [==============================] - 0s 4ms/step - loss: 0.1106 - accuracy: 0.9666 - val_loss: 0.0885 - val_accuracy: 0.9750
Epoch 8/10
45/45 [=============

The kernel size of my 2d convolution layer is 3x3 with 'relu' activation. My pooling layer is of size 2x2. I added two dense layers with different activations after flattening the convolved and pooled matrix. I trained it for 10 epochs and the final accuracy was 0.9833. The accuracy was relatively stable after epoch 6 so I don't think more training epochs are necessary.